src/Data/List/Types.purs

```diff
@@ -65,7 +64,47 @@ instance monoidList :: Monoid (List a) where
   mempty = Nil
 
 instance functorList :: Functor List where
-  map f = foldr (\x acc -> f x : acc) Nil
+  map = listMap
+
+-- chunked list Functor inspired by OCaml
+-- https://discuss.ocaml.org/t/a-new-list-map-that-is-both-stack-safe-and-fast/865
+-- chunk sizes determined through experimentation
+listMap :: forall a b. (a -> b) -> List a -> List b
+listMap f = startUnrolledMap unrollLimit
+  where
+  -- iterate the unrolled map up to 1000 times,
+  -- which hits up to 5000 elements
+  unrollLimit = 1000
+
+  startUnrolledMap :: Int -> List a -> List b
+  startUnrolledMap 0 (x : xs) = f x : chunkedRevMap xs
+  startUnrolledMap n (x1 : x2 : x3 : x4 : x5 : xs) =
```

Comment:
This can probably be made faster if you hand-optimize the cases. The JS generated here isn't going to be particularly optimal. We can reuse the branches.

Reply:
Yes, I thought the same thing. I'll look at it if the option you suggest with mutation isn't the way to go.

Reply:
Doesn't hand-optimizing mean that the code could break whenever the compiler chooses to lay out data differently?

Reply:
No, I just mean rewriting this so that common cases use the same branch, instead of having multiple branches which recognize lists of length >= 1 for example.

Reply:
If you look at the generated JS, you'll hopefully see what I mean.

Reply:
```purescript
case xs of
  x : xs' -> ...
    case xs' of
      x' : xs'' -> ...
      _ -> ...
  _ -> ...
```

PureScript doesn't perform this optimization, you have to do it manually.